File: .mpga-runtime/cli/src/commands/board-search.ts
import { findProjectRoot } from '../core/config.js';
import { loadAllTasks, type Task } from '../board/task.js';
import { log } from '../core/logger.js';
import path from 'path';

export interface BoardSearchOpts {
  priority?: string;
  column?: string;
  scope?: string;
  agent?: string;
  tags?: string;
}

/**
 * Search and filter board tasks by criteria.
 * Returns matching tasks (also prints them to console).
 */
export function handleBoardSearch(query: string, opts: BoardSearchOpts): Task[] {
  const projectRoot = findProjectRoot() ?? process.cwd();
  const tasksDir = path.join(projectRoot, 'MPGA', 'board', 'tasks');

  const allTasks = loadAllTasks(tasksDir);
  let results = allTasks;

  // Text search across task titles (case-insensitive)
  if (query && query.trim().length > 0) {
    const q = query.toLowerCase();
    results = results.filter((t) => t.title.toLowerCase().includes(q));
  }

  // Filter by priority
  if (opts.priority) {
    results = results.filter((t) => t.priority === opts.priority);
  }

  // Filter by column
  if (opts.column) {
    results = results.filter((t) => t.column === opts.column);
  }

  // Filter by scope
  if (opts.scope) {
    results = results.filter((t) => t.scopes.includes(opts.scope!));
  }

  // Filter by assigned agent
  if (opts.agent) {
    results = results.filter((t) => t.assigned === opts.agent);
  }

  // Filter by tags (comma-separated — task must have ALL specified tags)
  if (opts.tags) {
    const requiredTags = opts.tags.split(',').map((s) => s.trim());
    results = results.filter((t) => requiredTags.every((tag) => t.tags.includes(tag)));
  }

  // Print results
  if (results.length === 0) {
    log.info('No tasks match the given criteria.');
    return results;
  }

  log.header(`Search Results (${results.length} task${results.length === 1 ? '' : 's'})`);
  for (const t of results) {
    const parts = [
      t.id,
      t.title,
      `[${t.column}]`,
      t.priority,
    ];
    if (t.assigned) parts.push(`@${t.assigned}`);
    if (t.tags.length > 0) parts.push(`#${t.tags.join(',')}`);
    console.log(`  ${parts.join('  ')}`);
  }

  return results;
}

File: mpga-plugin/cli/src/mpga/commands/board_search.py
"""Search and filter board tasks by criteria."""

from __future__ import annotations

from pathlib import Path

from mpga.board.task import Task, load_all_tasks
from mpga.core.config import find_project_root
from mpga.core.logger import console, log


def handle_board_search(
    query: str,
    *,
    priority: str | None = None,
    column: str | None = None,
    scope: str | None = None,
    agent: str | None = None,
    tags: str | None = None,
) -> list[Task]:
    """Search and filter board tasks by criteria.

    Returns matching tasks (also prints them to console).
    """
    project_root = find_project_root() or Path.cwd()
    tasks_dir = str(project_root / "MPGA" / "board" / "tasks")

    all_tasks = load_all_tasks(tasks_dir)
    results = all_tasks

    # Text search across task titles (case-insensitive)
    if query and query.strip():
        q = query.lower()
        results = [t for t in results if q in t.title.lower()]

    # Filter by priority
    if priority:
        results = [t for t in results if t.priority == priority]

    # Filter by column
    if column:
        results = [t for t in results if t.column == column]

    # Filter by scope
    if scope:
        results = [t for t in results if scope in t.scopes]

    # Filter by assigned agent
    if agent:
        results = [t for t in results if t.assigned == agent]

    # Filter by tags (comma-separated -- task must have ALL specified tags)
    if tags:
        required_tags = [s.strip() for s in tags.split(",")]
        results = [t for t in results if all(tag in t.tags for tag in required_tags)]

    # Print results
    if not results:
        log.info("No tasks match the given criteria.")
        return results

    log.header(f"Search Results ({len(results)} task{'s' if len(results) != 1 else ''})")
    for t in results:
        parts = [t.id, t.title, f"[{t.column}]", t.priority]
        if t.assigned:
            parts.append(f"@{t.assigned}")
        if t.tags:
            parts.append(f"#{','.join(t.tags)}")
        console.print(f"  {'  '.join(parts)}")

    return results
